In [1]:
import pandas as pd

In [2]:
df_fusionné = pd.read_csv('df_fusionné.csv')

In [3]:
df_fusionné_binaire = df_fusionné.sample(frac=0.2)
df_fusionné_binaire.iloc[:, 0] = df_fusionné_binaire.iloc[:, 0].apply(lambda x: 0 if x == 'N' else 1)

In [4]:
# Count the occurrences of each class
counts = df_fusionné_binaire.iloc[:, 0].value_counts()

# Find the number of instances in the minority class
min_count = counts.min()

# Create balanced DataFrame through undersampling
df_balanced_under = pd.concat([
    df_fusionné_binaire[df_fusionné_binaire.iloc[:, 0] == 0].sample(min_count),
    df_fusionné_binaire[df_fusionné_binaire.iloc[:, 0] == 1].sample(min_count)
])

# Shuffle the DataFrame to mix the classes
df_balanced_under = df_balanced_under.sample(frac=1)


In [5]:
# Count occurrences of 0 and 1 in the first column
value_counts = df_balanced_under.iloc[:, 0].value_counts()

# Display the counts
print(value_counts)

ColumnName
1    7594
0    7594
Name: count, dtype: int64


In [6]:
df_balanced_under

ColumnName         0         1         2         3         4         5  \
10306           1  0.473727  0.473003  0.472581  0.472528  0.472420  0.472363   
79377           1  0.442109  0.442227  0.442360  0.442506  0.442654  0.442806   
87909           0  0.457614  0.457547  0.457487  0.457422  0.457330  0.457234   
59269           0  0.499488  0.502200  0.508321  0.514962  0.515455  0.513917   
83276           0  0.421630  0.419092  0.424596  0.411681  0.412613  0.414459   
...           ...       ...       ...       ...       ...       ...       ...   
29130           1  0.479325  0.477139  0.473028  0.478520  0.476689  0.474606   
185183          1  0.502949  0.520786  0.522212  0.519065  0.520744  0.523007   
155399          0  0.518201  0.513647  0.525294  0.508029  0.514353  0.522053   
166793          1  0.480011  0.465132  0.465820  0.471780  0.473930  0.475792   
189396          0  0.429810  0.428800  0.427864  0.427062  0.426341  0.425667   

               6         7         8  ...       710       711       712  \
10306   0.472604  0.472895  0.472716  ...  0.500201  0.500912  0.499307   
79377   0.442970  0.443138  0.443303  ...  0.441852  0.441881  0.441893   
87909   0.457124  0.457018  0.456977  ...  0.457978  0.457954  0.457922   
59269   0.511896  0.509571  0.509583  ...  0.663403  0.672461  0.647249   
83276   0.408233  0.400631  0.402279  ...  0.322368  0.325182  0.322451   
...          ...       ...       ...  ...       ...       ...       ...   
29130   0.476684  0.479893  0.480013  ...  0.543481  0.539800  0.539175   
185183  0.522358  0.521022  0.521855  ...  0.510629  0.511303  0.510981   
155399  0.518115  0.511143  0.512431  ...  0.416023  0.409964  0.397140   
166793  0.481365  0.487511  0.488762  ...  0.503300  0.501458  0.501840   
189396  0.425143  0.424653  0.423850  ...  0.445497  0.445141  0.444050   

             713       714       715       716       717       718       719  
10306   0.496620  0.493829  0.490669  0.487041  0.483392  0.479287  0.475414  
79377   0.441901  0.441918  0.441939  0.441952  0.441971  0.441989  0.442020  
87909   0.457888  0.457849  0.457808  0.457775  0.457742  0.457708  0.457670  
59269   0.610461  0.565314  0.519394  0.503355  0.499251  0.496536  0.497669  
83276   0.318273  0.319940  0.324675  0.322086  0.324533  0.332747  0.329460  
...          ...       ...       ...       ...       ...       ...       ...  
29130   0.538958  0.538728  0.538028  0.531784  0.520752  0.519351  0.527062  
185183  0.510082  0.510665  0.511013  0.508793  0.506979  0.499352  0.490820  
155399  0.382828  0.379605  0.382319  0.370250  0.366482  0.375528  0.367415  
166793  0.502561  0.503458  0.504086  0.500615  0.494928  0.493215  0.492629  
189396  0.442575  0.441031  0.439328  0.437391  0.435406  0.433249  0.431155  

[15188 rows x 721 columns]

In [7]:
import pandas as pd
from sklearn.model_selection import train_test_split

X = df_balanced_under.iloc[:, 1:]  
y = df_balanced_under.iloc[:, 0] 

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [8]:
print(y_train.unique())

[1 0]


In [9]:
y_train = y_train.astype(int)
y_test = y_test.astype(int)

In [10]:
import tensorflow as tf
from keras.models import Sequential
from keras.layers import Conv1D, BatchNormalization, MaxPooling1D, Dropout, Flatten, Dense
from keras.optimizers import Adam

Model = Sequential()

Model.add(Conv1D(512, 10, strides=1, padding="same", activation="relu", input_shape=(720, 1)))
Model.add(BatchNormalization())
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Conv1D(256, 10, strides=1, padding="same", activation="relu"))
Model.add(Dropout(0.3))
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Conv1D(128, 10, strides=1, padding="same", activation="relu"))
Model.add(Dropout(0.3))
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Conv1D(64, 10, strides=1, padding="same", activation="relu"))
Model.add(Dropout(0.3))
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Conv1D(32, 10, strides=1, padding="same", activation="relu"))
Model.add(Dropout(0.3))
Model.add(MaxPooling1D(3, strides=2, padding="same"))

Model.add(Flatten())
Model.add(Dense(1024, activation='relu'))
Model.add(Dropout(0.3))

Model.add(Dense(512, activation='relu'))
Model.add(Dropout(0.3))

Model.add(Dense(64, activation='relu'))
Model.add(Dropout(0.3))

Model.add(Dense(units=1, activation='sigmoid'))  # Use 'sigmoid' for binary classification


/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/keras/src/layers/convolutional/base_conv.py:99: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(
2024-03-13 17:45:02.233283: I metal_plugin/src/device/metal_device.cc:1154] Metal device set to: Apple M3 Pro
2024-03-13 17:45:02.233300: I metal_plugin/src/device/metal_device.cc:296] systemMemory: 18.00 GB
2024-03-13 17:45:02.233304: I metal_plugin/src/device/metal_device.cc:313] maxCacheSize: 6.00 GB
2024-03-13 17:45:02.233321: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:305] Could not identify NUMA node of platform GPU ID 0, defaulting to 0. Your kernel may not have been built with NUMA support.
2024-03-13 17:45:02.233331: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:271] Created TensorFlow

In [11]:
print(Model.summary())

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv1d (Conv1D)                 │ (None, 720, 512)       │         5,632 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ batch_normalization             │ (None, 720, 512)       │         2,048 │
│ (BatchNormalization)            │                        │               │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d (MaxPooling1D)    │ (None, 360, 512)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_1 (Conv1D)               │ (None, 360, 256)       │     1,310,976 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 360, 256)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_1 (MaxPooling1D)  │ (None, 180, 256)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_2 (Conv1D)               │ (None, 180, 128)       │       327,808 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 180, 128)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_2 (MaxPooling1D)  │ (None, 90, 128)        │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_3 (Conv1D)               │ (None, 90, 64)         │        81,984 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_2 (Dropout)             │ (None, 90, 64)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_3 (MaxPooling1D)  │ (None, 45, 64)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv1d_4 (Conv1D)               │ (None, 45, 32)         │        20,512 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_3 (Dropout)             │ (None, 45, 32)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling1d_4 (MaxPooling1D)  │ (None, 23, 32)         │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 736)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 1024)           │       754,688 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_4 (Dropout)             │ (None, 1024)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 512)            │       524,800 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_5 (Dropout)             │ (None, 512)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 64)             │        32,832 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_6 (Dropout)             │ (None, 64)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 1)              │            65 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 3,061,345 (11.68 MB)

 Trainable params: 3,060,321 (11.67 MB)

 Non-trainable params: 1,024 (4.00 KB)

None


In [12]:
# Define callbacks
Early_Stopper = tf.keras.callbacks.EarlyStopping(monitor="loss", patience=3, mode="min")


In [13]:
# Compile the model
Model.compile(optimizer=Adam(), loss='binary_crossentropy', metrics=['accuracy'])

In [14]:
Conv1D_Model = Model.fit(X_train, y_train, epochs=70, validation_data=(X_test, y_test), callbacks=[Early_Stopper])

Epoch 1/70


2024-03-13 17:45:03.008345: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:117] Plugin optimizer for device_type GPU is enabled.


380/380 ━━━━━━━━━━━━━━━━━━━━ 43s 110ms/step - accuracy: 0.6661 - loss: 0.6062 - val_accuracy: 0.6254 - val_loss: 0.6456
Epoch 2/70
380/380 ━━━━━━━━━━━━━━━━━━━━ 42s 110ms/step - accuracy: 0.7742 - loss: 0.4941 - val_accuracy: 0.6159 - val_loss: 0.6670
Epoch 3/70
380/380 ━━━━━━━━━━━━━━━━━━━━ 42s 111ms/step - accuracy: 0.7977 - loss: 0.4486 - val_accuracy: 0.5760 - val_loss: 0.7479
Epoch 4/70
380/380 ━━━━━━━━━━━━━━━━━━━━ 42s 111ms/step - accuracy: 0.8194 - loss: 0.4091 - val_accuracy: 0.7729 - val_loss: 0.4868
Epoch 5/70
380/380 ━━━━━━━━━━━━━━━━━━━━ 42s 112ms/step - accuracy: 0.8356 - loss: 0.3858 - val_accuracy: 0.6995 - val_loss: 0.5523
Epoch 6/70
380/380 ━━━━━━━━━━━━━━━━━━━━ 42s 111ms/step - accuracy: 0.8451 - loss: 0.3593 - val_accuracy: 0.6771 - val_loss: 0.5707
Epoch 7/70
380/380 ━━━━━━━━━━━━━━━━━━━━ 42s 112ms/step - accuracy: 0.8485 - loss: 0.3501 - val_accuracy: 0.6761 - val_loss: 0.6145
Epoch 8/70
380/380 ━━━━━━━━━━━━━━━━━━━━ 42s 112ms/step - accuracy: 0.8535 - loss: 0.3477 - val

In [15]:
Model.save('Conv1D_Model.h5')  # Saves the model to a HDF5 file

In [16]:
Model_Results = Model.evaluate(X_test,y_test)
print("LOSS:  " + "%.4f" % Model_Results[0])
print("ACCURACY:  " + "%.4f" % Model_Results[1])

95/95 ━━━━━━━━━━━━━━━━━━━━ 2s 17ms/step - accuracy: 0.8278 - loss: 0.3839
LOSS:  0.3770
ACCURACY:  0.8288


In [17]:
# Assuming your model is named 'Model'
y_pred_probs = Model.predict(X_test)
y_pred = (y_pred_probs > 0.5).astype("int32")  # Convert probabilities to binary labels


95/95 ━━━━━━━━━━━━━━━━━━━━ 2s 15ms/step


In [18]:
from sklearn.metrics import confusion_matrix, precision_score, recall_score, f1_score, roc_auc_score, roc_curve, log_loss

# Confusion Matrix
conf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:\n", conf_matrix)

# Precision
precision = precision_score(y_test, y_pred)
print(f"Precision: {precision:.2f}")

# Recall
recall = recall_score(y_test, y_pred)
print(f"Recall: {recall:.2f}")

# F1 Score
f1 = f1_score(y_test, y_pred)
print(f"F1 Score: {f1:.2f}")

# ROC-AUC Score
roc_auc = roc_auc_score(y_test, y_pred_probs)
print(f"ROC AUC Score: {roc_auc:.2f}")

# Now, you can calculate the log loss
logloss = log_loss(y_test, y_pred_probs)
print(f"Log Loss: {logloss:.2f}")

Confusion Matrix:
 [[1359  155]
 [ 365 1159]]
Precision: 0.88
Recall: 0.76
F1 Score: 0.82
ROC AUC Score: 0.92
Log Loss: 0.38


Confusion Matrix:
 [[1456  106]
 [ 640  834]]
Precision: 0.89
Recall: 0.57
F1 Score: 0.69
ROC AUC Score: 0.86
Log Loss: 0.65

Confusion Matrix:
 [[1357  127]
 [ 325 1187]]
Precision: 0.90
Recall: 0.79
F1 Score: 0.84
ROC AUC Score: 0.95
Log Loss: 0.33

Confusion Matrix:
 [[1401  104]
 [ 947  576]]
Precision: 0.85
Recall: 0.38
F1 Score: 0.52
ROC AUC Score: 0.81
Log Loss: 0.64


Confusion Matrix:
 [[1425  117]
 [ 375 1122]]
Precision: 0.91
Recall: 0.75
F1 Score: 0.82
ROC AUC Score: 0.93
Log Loss: 0.41
